File: pools/allocating_strategies/strategies/src/tests/unique_id.test.js
const strat = require('../unique_id_strategy')

test("unique id output and capacity", () => {
    let allocated = [uniqueId(0, 'first'),
        uniqueId(1, 'second'),
        uniqueId(3, 'third')];

    let resource_pool = {
        vpn: "VPN85", network: "Network19",
        idFormat: "VPN-{counter}-{network}-{vpn}-local"}

    let output = strat.invokeWithParams(allocated, resource_pool, {});
    expect(output)
        .toStrictEqual({'counter': 4, text: 'VPN-4-Network19-VPN85-local'});

    let capacity = strat.invokeWithParamsCapacity(allocated, resource_pool, {});
    expect(capacity)
        .toStrictEqual({freeCapacity: "9007199254740988", utilizedCapacity: "3"});
})

test("params without resourcePool", () => {
    let output = strat.invokeWithParams([], null, {});
    expect(output).toStrictEqual(null);
})

test("resourcePool without idFormat", () => {
    let output = strat.invokeWithParams([],
        {someProperty: "SomeUniqueL3VPN"}, {});
    expect(output).toStrictEqual(null)
})

test("idFormat without counter", () => {
    let output = strat.invokeWithParams([],
        {someProperty: "SomeUniqueL3VPN", idFormat: "{someProperty}"}, {});
    expect(output).toStrictEqual(null)
})

test("simple l3vpn counter", () => {
    let output = strat.invokeWithParams([],
        {someProperty: "L3VPN", idFormat: "{someProperty}{counter}"}, {});
    expect(output).toStrictEqual({'counter': 1, text: 'L3VPN1'});

    let capacity = strat.invokeWithParamsCapacity([uniqueId(output.counter, output.text)],
        {someProperty: "L3VPN", idFormat: "{someProperty}{counter}"}, {});
    expect(capacity).toStrictEqual({freeCapacity: "9007199254740990", utilizedCapacity: "1"});

    let next_output = strat.invokeWithParams([uniqueId(output.counter, output.text)],
        {someProperty: "L3VPN", idFormat: "{someProperty}{counter}"}, {});
    expect(next_output).toStrictEqual({'counter': 2, text: 'L3VPN2'});

    let next_capacity = strat.invokeWithParamsCapacity(
        [uniqueId(output.counter, output.text), uniqueId(next_output.counter, next_output.text)],
        {someProperty: "L3VPN", idFormat: "{someProperty}{counter}"}, {});
    expect(next_capacity).toStrictEqual({freeCapacity: "9007199254740989", utilizedCapacity: "2"});
})

test("multiple l3vpn counters", () => {
    let outputs = []
    for (let i = 1; i <= 10; i++) {
        let unique_id = strat.invokeWithParams(outputs,
            {someProperty: "L3VPN", idFormat: "{someProperty}{counter}"},
            {});
        outputs.push(uniqueId(unique_id.counter, unique_id.text))
        expect(unique_id).toStrictEqual({'counter': i, text: 'L3VPN'+ i })
    }

    let capacity = strat.invokeWithParamsCapacity(outputs,
        {someProperty: "L3VPN", idFormat: "{someProperty}{counter}"},
        {});
    expect(capacity).toStrictEqual({freeCapacity: "9007199254740981", utilizedCapacity: "10"});
})

test("simple range counter",() =>{
    let unique_id = strat.invokeWithParams([],
        {from: 1000, idFormat: "{counter}"},
        {});
    expect(unique_id).toStrictEqual({'counter': 1000, text: '1000' })
})

test("multiple range counter",() =>{
    let outputs = []
    for (let i = 1; i <= 10; i++) {
        let unique_id = strat.invokeWithParams(outputs,
            {from: 1000, idFormat: "{counter}"},
            {});
        outputs.push(uniqueId(unique_id.counter, unique_id.text))
        expect(unique_id).toStrictEqual({'counter': 999 + i, text: (999 + i).toString() })
    }
})
function uniqueId(counter, text) {
    return {"Properties": {"counter": counter, "text": text}}
}

File: pools/allocating_strategies/strategies/src/unique_id_test.go
package src

import (
	"reflect"
	"strconv"
	"testing"
)

func uniqueId(counter int, text string) map[string]interface{}{
	uniqueIdProperties := make(map[string]interface{})
	uniqueIdMap := make(map[string]interface{})
	uniqueIdMap["counter"] = counter
	uniqueIdMap["text"] = text
	uniqueIdProperties["Properties"] = uniqueIdMap
	return uniqueIdProperties
}

func TestUniqueIdOutputAndCapacity(t *testing.T) {
	var allocated = []map[string]interface{}{
		uniqueId(0, "first"),
		uniqueId(1, "second"),
		uniqueId(3, "third"),
	}

	var resourcePool = map[string]interface{}{
		"vpn": "VPN85", "network": "Network19",
		"idFormat": "VPN-{counter}-{network}-{vpn}-local",
	}
	uniqueIdStruct := NewUniqueId(allocated, resourcePool)

	var output = uniqueIdStruct.invoke()
	var expectedOutput = make(map[string]interface{})
	expectedOutput["counter"] = 4
	expectedOutput["text"] = "VPN-4-Network19-VPN85-local"

	eq := reflect.DeepEqual(output, expectedOutput)
	if !eq {
		t.Fatalf("different output of %s expected, got: %s", expectedOutput, output)
	}

	uniqueIdStruct = NewUniqueId(allocated, resourcePool)
	var capacity = uniqueIdStruct.capacity()
	var expectedCapacity = make(map[string]interface{})
	expectedCapacity["freeCapacity"] = int(^uint(0) >> 1) - 3  // Number.MAX_SAFE_INTEGER - 3
	expectedCapacity["utilizedCapacity"] = 3

	eq = reflect.DeepEqual(capacity, expectedCapacity)
	if !eq {
		t.Fatalf("different output of %s expected, got: %s", expectedCapacity, capacity)
	}
}

func TestSimpleRangeCounter(t *testing.T) {
	var outputs []map[string]interface{}
	var resourcePool = map[string]interface{}{"from": 1000, "idFormat": "{counter}"}
	uniqueIdStruct := NewUniqueId(outputs, resourcePool)
	for i := 0; i <= 10; i++ {
		var output = uniqueIdStruct.invoke()
		var expectedOutput = map[string]interface{}{"counter": 1000 + i, "text": strconv.Itoa(1000 + i)}
		eq := reflect.DeepEqual(output, expectedOutput)
		if !eq {
			t.Fatalf("different output of %s expected, got: %s", expectedOutput, output)
		}
		outputs = append(outputs, uniqueId(output["counter"].(int), output["text"].(string)))
		uniqueIdStruct = NewUniqueId(outputs, resourcePool)
	}

}

func TestParamsWithoutResourcePool(t *testing.T) {
	var allocated []map[string]interface{}

	uniqueIdStruct := NewUniqueId(allocated, nil)
	var output = uniqueIdStruct.invoke()

	eq := reflect.DeepEqual(output, (map[string]interface{})(nil))
	if !eq {
		t.Fatalf("different output of nil expected, got: %s", output)
	}
}

func TestResourcePoolWithoutIdFormat(t *testing.T) {
	var allocated []map[string]interface{}
	var resourcePool = map[string]interface{}{
		"vpn": "VPN85", "network": "Network19",
		"idFormat": "VPN-{network}-{vpn}-local",
	}
	uniqueIdStruct := NewUniqueId(allocated, resourcePool)
	var output = uniqueIdStruct.invoke()

	eq := reflect.DeepEqual(output, (map[string]interface{})(nil))
	if !eq {
		t.Fatalf("different output of nil expected, got: %s", output)
	}
}

func TestMultipleL3vpnCounters(t *testing.T) {
	var outputs []map[string]interface{}
	var resourcePool = map[string]interface{}{"someProperty": "L3VPN", "idFormat": "{someProperty}{counter}"}
	uniqueIdStruct := NewUniqueId(outputs, resourcePool)
	for i := 1; i <= 10; i++ {
		var output = uniqueIdStruct.invoke()
		var expectedOutput = map[string]interface{}{"counter": i, "text": "L3VPN" + strconv.Itoa(i)}
		eq := reflect.DeepEqual(output, expectedOutput)
		if !eq {
			t.Fatalf("different output of %s expected, got: %s", expectedOutput, output)
		}
		outputs = append(outputs, uniqueId(output["counter"].(int), output["text"].(string)))
		uniqueIdStruct = NewUniqueId(outputs, resourcePool)

		var capacity = uniqueIdStruct.capacity()
		var expectedCapacity = make(map[string]interface{})
		expectedCapacity["freeCapacity"] = int(^uint(0) >> 1) - i  // Number.MAX_SAFE_INTEGER - 3
		expectedCapacity["utilizedCapacity"] = i
		eq = reflect.DeepEqual(capacity, expectedCapacity)
		if !eq {
			t.Fatalf("different output of %s expected, got: %s", expectedCapacity, capacity)
		}
	}
}